viewer/viewer.js

```diff
@@ -274,6 +275,27 @@ if (Config.get("passwordSecret")) {
     if (internals.userNameHeader !== undefined) {
       if (req.headers[internals.userNameHeader] !== undefined) {
         var userName = req.headers[internals.userNameHeader];
+        // Auto-creation of users from headers
+        if (internals.userAutoCreateTmpl !== undefined) {
+
+           // Check to see if we need to auto-create.
+           Db.getUserCache(userName, function(err, suser) {
+              if (err) {return res.send("ERROR - getUser - user: " + userName + " err:" + err);}
+              if (!suser || !suser.found) {
+                 var nuser = JSON.parse(new Function("return `" + internals.userAutoCreateTmpl + "`;").call(req.headers));
+                 var res = Db.setUser(userName, nuser, (err, info) => {
+                   if (err) {
+                     console.log("Elastic search error adding user: (" +  userName + "):(" + JSON.stringify(nuser) + "):" + err);
+                   } else {
+                     console.log("Added user:" + userName + ":" + JSON.stringify(nuser));
+                   }
+                 });
+              }
+              return;
```

Comment:
You need to actually call the userCleanup/next here too, since because of race conditions it might not hit the one below.  Actually I would probably switch the order around where you do the getUserCache always and then have the if internals.userAutoCreateTmpl !== undefined inside the results if the user isn't found.

Reply:
Agreed, though wouldn't i need to call getUserCache again recursively in the cb from inside getUserCache?  Is there a good way to avoid code duplication (I'm admittedly new to javascript)

Reply:
I might not be following your question but the issue with the current code is when the user doesn't exist in cache its going to call ES twice at the same time to get the user, and the second one will fail, before the first one can create it.

What you have now is
```
if () {
  getUserCache() {  #1
    createUser
    return
  }
}
getUserCache () { #2
  //the user won't exit here
}

Reply:
oh, i see.  I will fix that...  I didn't realize that would go async like that...  i have some node learning to do i suppose

Reply:
I rewrote the algorithm now that i understand the `next()` concept.  Happy to make more modifications.